clusterloader: enhance namespace configurations

yuanchen8911 opened the pull request:
This PR is to separate namespace specific configurations from cluster level configurations (e.g., kubeconfig, master-ip). The cluster configurations belong to config.ClusterConfig and are specified as the command line options while namespace configuration is a part of api.Config and deigned in the testconfig.yaml file.

1.  A new structure _NamespaceConfig_ is defined for namespace parameters as a part of api.Config and the values are specified in the testconfig yaml file.  If a parameter is not specified, a default value is used.
2. A new parameter "_Prefix_" is added to specify the prefix of auto-created namespaces. Clusterloader uses the specified prefix if set otherwise uses the default "_test-\<random string\>_".
3. Add two files _api.default.go_ and _api.validate.go_  for setting default parameters and checking the validness of namespace parameters. 
4. _"automanagedNamespaces"_  in testconfig file and _"delete-stale-namespaces"_ and _"delete-automananged-namespaces"_ in the command line can still be used with deprecate warnings. 

New structure for namespace configuration
```
// NamespaceConfig defines namespace parameters.
type NamespaceConfig struct {
        Number int32  `json: number` 
        Prefix string  `json: prefix`
        DeleteStaleNamespaces *bool `json.deleteStaleNamespaces`
        DeleteAutomanagedNamespaces *bool `json.deleteAutomanagedNamespaces`
}
```
An example namespace configuration in the testconfig file.
```
namespace:
   number: 1
   namePrefix: "test"
   EnableExistingNamespaces: false
   DeleteStaleNamespaces: false
```
An issue https://github.com/kubernetes/perf-tests/issues/1169 was opened and discussed with @wojtek-t. The support of pre-created existing namespaces will be another PR after this PR is approved and merged.

yuanchen8911 commented:
/assign @mm4tt

wojtek-t commented on clusterloader2/testing/access-tokens/config.yaml:
```diff
@@ -42,7 +42,11 @@
+  count: {{$namespaces}}
+  prefix:
+  deleteStaleNamespaces: false
+  deleteAutomanagedNamespaces: true
```
I really don't like the fact we need to set those values in every test - we need to smartly default them instead and rely on the defaulting whenever it is possible.

yuanchen8911 replied:
OK, will set the following default values. 
```
count=1
prefix= ""
deleteStaleNamespaces=false
deleteAutomanagedNamespace=true
```

yuanchen8911 replied:
removed

wojtek-t commented on clusterloader2/api/types.go:
```diff
@@ -41,8 +41,8 @@ type TestScenario struct {
-	// AutomanagedNamespaces is a number of automanaged namespaces.
-	AutomanagedNamespaces int32 `json: automanagedNamespaces`
+	// Namespace is a structure for namespace configuration.
+	Namespace *NamespaceConfig `json: namespace`
```
We can't do that in a single step. We should:
- add this field and mark the AutomanagedNamespaces field as deprecated (though still respect it somehow)
- allow people for doing the transition
- only in some time get rid of this field

yuanchen8911 replied:
sure,  we'll keep it as as deprecated.

yuanchen8911 replied:
The following default values are used if a parameter is not specified in either new or old way. 
```
     count: 1
     prefix: ""
     deleteStaleNamespaces: false
     deleteAutomananagedNamespaces: true
```

yuanchen8911 replied:
AutomanangedNamespaces is kept.  If the number of namespaces is specified by both AutoMananagedNamespaces and the new structure, the value of new structure will be used. If neither defines the parameter, a default value of 1 will be used.  The new code also checks the value and reports an error if it's less than 1.

wojtek-t commented on clusterloader2/api/types.go:
```diff
@@ -106,6 +106,22 @@ type ListUnknownObjectOptions struct {
+	DeleteAutomanagedNamespaces bool `json: deleteAutomanagedNamespaces`
+	// EnableExistingNamespace specifies whether to use existing namespaces
+	// if they exist  (TO DO)
+	EnableExistingNamespaces bool `json: enableExistingNamespaces`
```
Let's not add fields that are not yet supported.

yuanchen8911 replied:
will delete it.

yuanchen8911 replied:
deleted it.

wojtek-t commented on clusterloader2/cmd/clusterloader.go:
```diff
@@ -66,8 +66,6 @@ func initClusterFlags() {
 	flags.StringEnvVar(&clusterLoaderConfig.ClusterConfig.EtcdCertificatePath, "etcd-certificate", "ETCD_CERTIFICATE", "/etc/srv/kubernetes/pki/etcd-apiserver-server.crt", "Path to the etcd certificate on the master machine")
 	flags.StringEnvVar(&clusterLoaderConfig.ClusterConfig.EtcdKeyPath, "etcd-key", "ETCD_KEY", "/etc/srv/kubernetes/pki/etcd-apiserver-server.key", "Path to the etcd key on the master machine")
 	flags.IntEnvVar(&clusterLoaderConfig.ClusterConfig.EtcdInsecurePort, "etcd-insecure-port", "ETCD_INSECURE_PORT", 2382, "Inscure http port")
-	flags.BoolEnvVar(&clusterLoaderConfig.ClusterConfig.DeleteStaleNamespaces, "delete-stale-namespaces", "DELETE_STALE_NAMESPACES", false, "Whether to delete all stale namespaces before the test execution.")
```
I'm probably fine with deleting those. But it would be great if we could provide a defaulting in a way that would accept the CL2_DELETE_STALE_NAMESPACES env var as overwrite for it.
Didin't think about it much - @mm4tt - do you think it would be possible?

yuanchen8911 replied:
To maintain back compatibility,  it might be good to keep it for the time being. I'll undo the deletion.

yuanchen8911 replied:
To maintain backward compatibility, the current command line options and environment variables "delete-stale-namespaces" and "delete-automanaged-namespaces" are kept with deprecated warnings.  automananagedNamespaces can still be used to specify the number of namespaces in testconfig file. If a parameter is not specified, a default value is used. If a parameter is specified in both new and old ways, the value using the new way is used.

yuanchen8911 commented:
@wojtek-t Thanks for the very helpful feedback! Here is a summary of the updated version. I hope it addressed most of your comments and concerns. 

1. **Backward compatibility**: the use of command line options (and environment variables) to specify "_delete-stale-namespace_" and "_delete-automanaged-namespaces_" and the use of automanagedNamespaces to define the number of namespaces in testconfig file can still be used with deprecated warnings or flag. So scripts and configuration files for the current version should continue to work without any changes.

2. **Default values**: defined the following default value if a parameter is not defined either way.
```
    count: 1
     prefix: ""
     deleteStaleNamespaces: false
     deleteAutomananagedNamespaces: true
```

3. If a parameter is specified in both old and new ways, the value defined using the new way is used.

4. Check the value of automanagedNamespaces and report an error if the specified value for it is less than 1.

yuanchen8911 commented:
/assign @wojtek-t 
/assign @krzysied

wojtek-t commented on clusterloader2/api/types.go:
```diff
@@ -106,6 +109,19 @@ type ListUnknownObjectOptions struct {
+// NamespaceConfig defines namespace parameters.
+type NamespaceConfig struct {
+	// Count is a number of automanaged namespaces.
+	Count *int32 `json: count,omitempty`
```
Let's remove the pointer here.
Basically, "0" is the bad value (the test doesn't make sense for count=0). So we don't really need to distinguish it from nil - we can simply default if set to 0.

wojtek-t commented on clusterloader2/pkg/test/simple_test_executor.go:
```diff
@@ -318,6 +320,52 @@ func (ste *simpleTestExecutor) ExecuteObject(ctx Context, object *api.Object, na
+	}
+	if *conf.Namespace.Count <= 0 {
+		return errors.NewErrorList(fmt.Errorf("number of namespaecs: %d was less than 1", *conf.Namespace.Count))
+	}
```
This technically has nothing to do with executor.
So far we never have a need for it, but I would prefer to:
- in api/ directory introduce the default.go file
- in that file define functions like:
```
func (c *Config) SetDefaults() {
  c.Namespace.SetDefaults()
}

func (n *NamespaceConfig) SetDefaults() {
  // defaulting logic that you have here
}
```

On top of that, introduce a `validation.go` where you would introduce define:
```
func (c *Config) Validate() errors.ErrorList {
  return c.Namespace.Validate()
}

func (n *NamespaceConfig) Validate() errors.ErrorList {
  // return what you have here
}
```

And call defaulting and validation here:
https://github.com/kubernetes/perf-tests/blob/d9bdf46ed8414e87b3c50d49f14a64e9e8970051/clusterloader2/pkg/test/test.go#L63

yuanchen8911 replied:
> This technically has nothing to do with executor.
> So far we never have a need for it, but I would prefer to:
> 
> * in api/ directory introduce the default.go file
> * in that file define functions like:
> 
> ```
> func (c *Config) SetDefaults() {
>   c.Namespace.SetDefaults()
> }
> 
> func (n *NamespaceConfig) SetDefaults() {
>   // defaulting logic that you have here
> }
> ```
> 
> On top of that, introduce a `validation.go` where you would introduce define:
> 
> ```
> func (c *Config) Validate() errors.ErrorList {
>   return c.Namespace.Validate()
> }
> 
> func (n *NamespaceConfig) Validate() errors.ErrorList {
>   // return what you have here
> }
> ```
> 
This would certainly lead to a neater implementations, but I have two comments/questions. 

1. Supporting the deprecated command line options _delete-stale-namespaces_ would be difficult because SetDefaults does not have access to config.ClusterConfig.DeleteStaleNamespaces. I'd suggest we just delete this command flag.

2.  To continue support automanagedNamespaces in the testconfig file would be tricky too. Without using *int32, when Count's value is 1 (the default), it's impossible to tell if it's explicitly set or the default value is used. In practice, a user should only specify the number in one or the other way, but not both.  In case both are specified (not equal to 1), we'll use the Count's value.
```
@wojtek-t

yuanchen8911 replied:
done

wojtek-t commented on clusterloader2/pkg/test/simple_test_executor.go:
```diff
@@ -318,6 +320,52 @@ func (ste *simpleTestExecutor) ExecuteObject(ctx Context, object *api.Object, na
+	}
+
+	// For backwarad compatibility with deprecated options "delete-stale-namespaces" and "delete-automananged-namespaces".
+	// Remove in the future versions
```
We don't have any formal deprecation policy, so let's remove it (it's not perfect, but it will be much cleaner).

yuanchen8911 replied:
done,  new implementation in api.default.go and api.validate.go as suggested.

wojtek-t commented on clusterloader2/testing/access-tokens/config.yaml:
```diff
@@ -42,7 +42,11 @@
-automanagedNamespaces: {{$namespaces}}
+namespace:
+  count: {{$namespaces}}
+  prefix:
```
With the defaulting you added, neither of those below should be needed (you only need to specify count).

yuanchen8911 replied:
done

yuanchen8911 commented:
@wojtek-t Based on your feedback,  I've made the following changes. 

1. Added two files default.go and validate.go for setting the default values and checking the validness of the parameters for namespaces respectively.
2. Use _int32_ and _bool_  instead of _*int32_ and _*bool_ in NamespaceConfig.
3. To maintain the compatibility,  _"automanagedNamespaces"_ can still be used to specify the number of namespaces in testconfig file, but the use of command line options to specify " _delete-stale-namespaces_" and "_delete-automanaged-namespaces_" are removed  because it's hard to implement it in a neat way.  If you have a good idea, please let me know.
4. Removed additional namespace parameters in config*.yaml as you suggested.

yuanchen8911 commented:
/assign @wojtek-t

yuanchen8911 commented:
/assign @krzysied

wojtek-t commented on clusterloader2/api/default.go:
```diff
@@ -0,0 +1,38 @@
+
+// SetDefaults set the default configuration parameters.
+func (conf *Config) SetDefaults() {
+	//To support deprecated "automanangedNamespaces" in testconfig
```
Let's remove this comment.

wojtek-t commented on clusterloader2/api/default.go:
```diff
@@ -0,0 +1,38 @@
+// SetDefaults set the default configuration parameters.
+func (conf *Config) SetDefaults() {
+	//To support deprecated "automanangedNamespaces" in testconfig
+	conf.AutomanagedNamespaces = 1
```
I don't think it's needed. Below we default `namespace.number`, so that one will be used anyway..
I think we should remove it.

yuanchen8911 replied:
done

wojtek-t commented on clusterloader2/api/default.go:
```diff
@@ -0,0 +1,38 @@
+
+// SetDefaults specifies the default values for namespace parameters.
+func (ns *NamespaceConfig) SetDefaults() {
+	ns.Number = 1
```
That is incorrect. You should default to 1, only if it was previously set to 0.

yuanchen8911 replied:
@wojtek-t  I'm not sure about this one. My implementation is to set the defaults before parsing the testconfig yaml file. The default value will be overridden if it's set in testconfig.
I think it would be a neater implementation. If we set the defaults after unmarshalling the testconfig, we have to check if "0" or false is set or unset using pointers.  The same to the delete namespace flags. I already tested it with different cases and worked as expected. 

https://github.com/yuanchen8911/perf-tests/blob/master/clusterloader2/pkg/test/test.go
```
# Directory Structure

- profile: clusterloader template files for different test cases
- profile/override: test case-specific parameter files
- script: scirpts to run tests and manage kubernetes resources
- config: kubelets config files
```

wojtek-t replied:
I comment on that too - I'm not a fan of this approach. I much more prefer the more standard approach (and what k8s itself is doing internally), which is what I suggested:
- read the config
- for fields that are unset, propagate with defaults

yuanchen8911 replied:
My current implementation is to set the defaults first and override them if set later.

yuanchen8911 replied:
OK, I'll make the changes.

wojtek-t commented on clusterloader2/api/default.go:
```diff
@@ -0,0 +1,38 @@
+// SetDefaults specifies the default values for namespace parameters.
+func (ns *NamespaceConfig) SetDefaults() {
+	ns.Number = 1
+	ns.Prefix = fmt.Sprintf("test-%s", util.RandomDNS1123String(6))
```
This also is incorrect - this should be conditional

yuanchen8911 replied:
Same as the above comment, the defaults are set first. If a filed is specified in the yaml file and the default value will be overridden.

wojtek-t commented on clusterloader2/api/default.go:
```diff
@@ -0,0 +1,38 @@
+func (ns *NamespaceConfig) SetDefaults() {
+	ns.Number = 1
+	ns.Prefix = fmt.Sprintf("test-%s", util.RandomDNS1123String(6))
+	ns.DeleteStaleNamespaces = false
```
Same here. And this has to be a pointer, because otherwise you have no way to distinguish between unset value and explicitly set to false.
[It's also a reason  why I previously commented about removing a pointer only for Number field.]

yuanchen8911 replied:
Same here. The defaults are set using the command line flags or default values before parsing the yaml file and will be overridden by the values in testconfig file. This is an simple and clean implementation in my opinion.

yuanchen8911 replied:
made the change

wojtek-t commented on clusterloader2/api/types.go:
```diff
@@ -106,6 +109,19 @@ type ListUnknownObjectOptions struct {
+	// It's optional, if set CL will use it, otherwise generate one with random string.
+	Prefix string `json: prefix,omitempty`
+	// DeleteStaleNamespaces specifies whether or not delete stale namepaces
+	DeleteStaleNamespaces bool `json: deleteStaleNamespaces,omitempty`
```
As mentioned above - these bool fields should be pointers.

yuanchen8911 replied:
Changed

wojtek-t commented on clusterloader2/api/validation.go:
```diff
@@ -0,0 +1,54 @@
+
+	// To support deprecated automanagedNamespaces.
+	if conf.AutomanagedNamespaces > 1 {
+		conf.Namespace.Number = conf.AutomanagedNamespaces
```
Validation should not change anything.

yuanchen8911 replied:
Changed it.

wojtek-t commented on clusterloader2/pkg/config/codec.go:
```diff
@@ -35,12 +35,11 @@ var (
-	var config api.Config
-	if err := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(raw), 4096).Decode(&config); err != nil {
-		return nil, fmt.Errorf("decoding failed: %v", err)
+func convertToConfig(config *api.Config, raw []byte) error {
```
Why did you change this?

yuanchen8911 replied:
@wojtek-t  Because the default values are set before parsing the testcong yaml file. So  I'll need to define testConfig in test/test.go and pass the pointer to TemplateToConfig in config/template_provider.go. The current version is to define testConfig in config/codec.go and return the pointer to it.

yuanchen8911 replied:
Changed it to parse confit first and then set defaults for unset Parma.

wojtek-t commented on clusterloader2/pkg/test/simple_test_executor.go:
```diff
@@ -52,9 +52,11 @@ func createSimpleTestExecutor() TestExecutor {
 // ExecuteTest executes test based on provided configuration.
 func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) *errors.ErrorList {
-	ctx.GetClusterFramework().SetAutomanagedNamespacePrefix(fmt.Sprintf("test-%s", util.RandomDNS1123String(6)))
+
```
remove this empty line

yuanchen8911 replied:
Deleted it

wojtek-t commented on clusterloader2/pkg/test/test.go:
```diff
@@ -57,9 +58,23 @@ func RunTest(clusterFramework, prometheusFramework *framework.Framework, cluster
+
+	// Set the default values for configuration.
+	var testConfig api.Config
+	testConfig.SetDefaults()
```
I'm not a fan of this pattern. You should leave what was here as it was and simply do:
```
testConfig.SetDetauls()
if err := testConfig.Validate(); err != nil {
  // ....
}
```

yuanchen8911 replied:
The implementation logic is 

1. set defaults (1 for namespaces,   delete ns flags from command line defaults or settings
2. decode yaml test config
3.  verify the validness of the values

So testConfig needs to be defined  first.

wojtek-t replied:
As mentioned in: https://github.com/kubernetes/perf-tests/pull/1173#discussion_r410413673
I don't think what you did is what we really want. We don't want to come up with different patterns for things that were already solved in Kubernetes itself.

yuanchen8911 replied:
changed as suggested. Please review. Thanks.

wojtek-t commented on clusterloader2/pkg/test/test.go:
```diff
@@ -57,9 +58,23 @@ func RunTest(clusterFramework, prometheusFramework *framework.Framework, cluster
+	// Set the default values for configuration.
+	var testConfig api.Config
+	testConfig.SetDefaults()
+	// Support the deprecated command line options
```
please add a TODO to remove them.

yuanchen8911 replied:
Done

yuanchen8911 commented:
@wojtek-t  OK, I'll make the changes by decoding the yaml file first and setting the defaults if not set.

yuanchen8911 commented:
@wojtek-t The implementation logic was modified as you suggested. Please review. Thanks!
1. Parse testconfig
2. Set defaults for unset parameters

To support current command line options delete-stale-namepsace and delete-automanaged-namespaces, their values are as the default values before call SetDefault. Add "TODO:  remove them in the future versions."

wojtek-t commented on clusterloader2/api/validation.go:
```diff
@@ -0,0 +1,36 @@
+// Validate checks and verifies the namespace parameters.
+func (ns *NamespaceConfig) Validate() *errors.ErrorList {
+	if ns.Number <= 0 {
+		return errors.NewErrorList(fmt.Errorf("number of namespaces: %d was less than 1", ns.Number))
```
We would like to return all errors. I know in the current form it doesn't matter. But for future extensibility, let's change:
```
result := errors.NewErrorList()
if ns.Number <= 0 {
   // Append
}

if result.IsEmpty() {
  return nil
}
return result
}
```

yuanchen8911 replied:
@wojtek-t  Thank, the following change was made as suggested. 

```
        errList := errors.NewErrorList()
        if ns.Number <= 0 {
                errList.Append(fmt.Errorf("number of namespaces: %d was less than 1", ns.Number))
        }

        if errList.IsEmpty() {
                return nil
        }

        return errList

```

wojtek-t commented:
/lgtm
/approve

Thanks!
